staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -66,28 +69,61 @@ func newWithClock(authenticator authenticator.Token, cacheErrs bool, successTTL,
 	}
 }
 
+const (
+	// We have to cache for a non-zero amount of time in order to be able to share work.
+	minimumCacheTTL = 1 * time.Microsecond
+
+	// if it takes longer than this to perform the underlying auth call, it
+	// will still succeed, but other calls arriving after this time and
+	// before its completion will not share its result.
+	maxComputationTimeForJoining = 500 * time.Millisecond
+)
+
+func (a *cachedTokenAuthenticator) lookupFunc(ctx context.Context, token string) lrucache.ComputeFunc {
+	return func(abort <-chan time.Time) (value interface{}, ttl time.Duration) {
+		ctx, cancel := context.WithCancel(ctx)
+		defer cancel()
+		go func() {
+			select {
+			case <-abort:
+				cancel()
+			case ctx.Done():
+			}
+		}()
+		resp, ok, err := a.authenticator.AuthenticateToken(ctx, token)
+		if !a.cacheErrs && err != nil {
+			// caching the error for the briefest amount of time is
+			// how we let all waiting on the result see it.
+			return &cacheRecord{resp: resp, ok: ok, err: err}, minimumCacheTTL
```

Comment:
@liggitt @mikedanese If I cache uncacheable errors for 1us, is that going to be a problem?

Joining up work means that concurrent requests within 500ms share fate anyway.

I can't avoid sharing this because I don't know ahead of time if there's going to be an error.

Sound OK?

Reply:
possibly. would like to take a look before this merges. should be today/tomorrow.